tools/calyx-pass/src/main.rs

```diff
@@ -0,0 +1,283 @@
+    // to target/debug/calyx
+    if args.calyx_exec.is_empty() {
+        args.calyx_exec = "target/debug/calyx".into();
+
+        if let Ok(calyx_exec_rel) = util::capture_command_stdout(
+            "fud",
+            &["config", "stages.calyx.exec"],
+            true,
+        ) {
+            args.calyx_exec = calyx_exec_rel.trim().into();
+        }
+    }
+
+    if !args.disable.is_empty() && args.breakpoint.is_none() {
+        fail("Invalid command line flags.", || {
+            println!("Using the disable pass option (`-d`) requires a breakpoint to be set. You can set one with `-b`.");
+        });
+    }
+
+    assert!(!args.calyx_exec.is_empty());
+
+    match util::capture_command_stdout(&args.calyx_exec, &["--version"], true) {
+        Ok(_) => {}
+        Err(_) => {
+            fail("Failed to determine or repair calyx executable path automatically.", || {
+                println!("{}", "Here's how to fix it:".bold());
+                println!("Option 1. Setup your fud config so that 'stages.calyx.exec' yields a valid path to the calyx executable");
+                println!("Option 2. Determine the path manually and pass it to the `-e` or `--calyx-exec` option");
+                println!("Option 3. Run this tool from the repository directory after calling `cargo build`");
+            });
+        }
+    }
+
+    if args.input_file.is_none() {
+        fail("Invalid command line arguments.", || {
+            println!("You must pass a single calyx program as input. However, when the version is requested through `--version`, this input file is not required and will be ignored.");
+        });
+    }
+
+    // use . for tmpdir for debugging, eventually just use TempDir::new
+    let temp_dir = TempDir::new_in(".", ".calyx-pass")?;
+    let mut pass_explorer = PassExplorer::new(
+        temp_dir,
+        args.calyx_exec,
+        args.breakpoint
+            .map(|pass| Breakpoint::from(pass, args.disable)),
+        args.pass_alias,
+        PathBuf::from(
+            args.input_file.expect("No input file passed as required"),
+        ),
+    )?;
+
+    /// Quit the program.
+    const QUIT: char = 'q';
+
+    /// See [`PassExplorer::accept`].
+    const ACCEPT: char = 'a';
+
+    /// See [`PassExplorer::skip`].
+    const SKIP: char = 's';
+
+    /// See [`PassExplorer::undo`].
+    const UNDO: char = 'u';
+
+    /// Scroll forward [`JUMP`] lines.
+    const JUMP_FWD: char = 'f';
+
+    /// Scroll backward [`JUMP`] lines.
+    const JUMP_BCK: char = 'b';
+
+    /// See [`JUMP_FWD`] and [`JUMP_BCK`].
+    const JUMP: usize = 4;
+
+    let mut term_stdout = Term::stdout();
+    writeln!(term_stdout, "{}", SAVE_SCREEN)?;
+    writeln!(term_stdout, "{}", SAVE_CURSOR)?;
+    writeln!(term_stdout, "{}", SWITCH_ALTERNATIVE_BUFFER)?;
+    term_stdout.hide_cursor()?;
+
+    let mut scrollback_buffer = ScrollbackBuffer::new(&term_stdout);
+    let mut needs_redraw = true;
+
+    loop {
+        if needs_redraw {
```

Comment:
This nest should be broken up into different functions. Maybe between `draw` and `next_task` or something

Reply:
<img width="527" alt="image" src="https://github.com/calyxir/calyx/assets/113849268/ffb85a26-8d6e-42a2-b684-c768a208f7a7">